homeassistant/components/heos/media_player.py

```diff
@@ -167,9 +169,49 @@ def __init__(self, player):
         """Play a piece of media."""
         if media_type == MEDIA_TYPE_URL:
             await self._player.play_url(media_id)
-        else:
-            _LOGGER.error("Unable to play media: Unsupported media type '%s'",
-                          media_type)
+            return
+
+        if media_type == "quick_select":
+            # media_id may be an int or a str
+            selects = await self._player.get_quick_selects()
+            try:
+                index = int(media_id)
+            except ValueError:
+                # Try finding index by name
+                index = next((index for index, select in selects.items()
+                              if select == media_id), None)
+            if index is None:
+                raise ValueError("Invalid quick select '{}'".format(media_id))
```

Comment:
I agree with @elupus. We shouldn't raise standard library errors raw. Create a new `HomeAssistantError` and raise that and then catch that in the decorator.

Reply:
I'm not sure what the reference is to @elupus.  For my own learning, why is using `HomeAssistantError` better in this case?  Since we're just going to catch it in the decorator and only log the message (not the exception type), it won't make any difference.  Also, we'll still have to catch `ValueError` in the decorator because the library will raise this if the index is invalid (i.e. outside of the valid range).

Reply:
Well in general, the point is to avoid hiding errors. You want to catch as specific as possible and let unknown error propagate up. That way it is much easier to debug faults and not have silent fails. If something else now unexpectedly raises value error, you will catch that and act potentially invalidly on that.

Also, a library should wrap known exceptions in their own exceptions to avoid dependency leaking. So a library using internally the request lib, for example, to make http requests should never expose error from the request lib in their raw form and expect user of lib to catch them. Say the library suddenly switches to some new cool http reader throwing other exceptions. Now users of lib need to rewrite since the leaked exceptions have changed. It's a abstraction leak.

If library indeed does raise value error, you should try to catch that close to the api call, where you know the likely source and if needed reraise as a know specific exception type.

That said. I don't think any of this is blocking this pull. It can imho be merged with this behavior now and fixed at a later date.

Reply:
An additional note on wrapping. Its usually good to keep original exception around when catching and raising using raise LibspecificException() from original_exception. That way its still possible to handle original cause.

Reply:
Note, this PR has already been merged.  This is good, thanks for the feedback.  I'll make those recommend changes to the library, as it's certainly leaking implementation now, and will make this code easier to maintain in HA.

What is your perspective for this use case: The API in the library accepts a parameter that it expects to be an `int > 0`.  Should it not raise `ValueError` if the `int` is `-1`?  That's validation, not leaking the implementation.  I'm used to other languages where it was best practice to use the built-in exceptions (like `ArgumentException` in C# for cases like this) for parameter guarding.  What's your recommendation in this case?  Is `ValueError` appropriate or should that also be wrapped in a custom exception?

Reply:
I agree with you here. ValueError and ArgumentException are builtins to Python so for that use case it should be fine to raise them. But those types of exceptions can't be allowed to propage too far since then the context is lost. And it's not leaking anything in that case.

Reply:
Forgot to mention, you can raise a:

```
class LibExceptionValue(ValueError):
  pass
```

Ie one that inherits from ValueError.